### DISCUSSION TITLE
Musings on Asset pipline

### DISCUSSION THREAD

**Discussion by nemerle** [o3de/o3de#4350] [Ideas]
I've been looking over the o3de's asset pipeline and I have trouble finding information about the scenarios described below.

Very likely they are 'there' but I can't find them, and what little documentation is there, is not helping all that much :smile:  

Definitions: 
* Data Source - a local file, a remote file, a rich-socket ( at least with a notion of a data version to support correct syncing )
* Source Asset (SA) - a data source that is processed by AssetProcessor to produce IntermediateAsset/s
* Intermediate Asset (IA) - a data source that can easily be accessed by various engine tools (editor/game/server)
* Final Asset (FA) - platform-specific data source that is optimized for release, built during the bundling process

Use scenarios :
 
* Defining asset import pipeline parameters:  ( SA->IA ) ( pipeline config default/per asset ): 
  *  Image format conversion ( texture -> basis etc., color processing )
  *  Audio format conversion ( target format selection from a minimal set [wav/ogg]?, ambisonics? )
  *  Mesh processing ( optimization, buffer preparation, LOD generation )
  *  Scene import ( procedural processing, prefab identification - finding commonly used sub-trees, etc. )
  *  Custom asset pipelines ( convert csv into textured meshes, other fun stuffs :) )
 

* Defining asset finalization pipeline parameters (IA->FA ) ( pipeline config default/per platform/per asset ): 
  * Texture downscaling ( format selection, number of mips )
  * Movie resizing/recoding
  * Custom asset packing algorithms ( animation compression, mesh compression, etc. )

* DCC live-link ( https://github.com/unity3d-jp/MeshSync )
  * We will likely want two-way interaction, editing an asset in the editor should be reflected back to the DCC tool.

 
Do we even have a notion of Intermediate Asset?
How can we support archives ( maybe even nested ones ) as Source Assets?

**Comment by Infurael** [comment]
"Do we even have a notion of Intermediate Asset?"
These are supported, in a way. Intermediate assets can be exported to the same cache that final assets are written to. Intermediate assets can in turn cause new jobs to queued on the Asset Processor. Assets can report dependencies which describes the relation individual files have. The Asset Bundler can then take that information to get just the files the runtime needs and ignored the intermediate assets.

"How can we support archives ( maybe even nested ones ) as Source Assets?"
Individual Asset Builders can do the decompression locally, but if multiple source assets are involved than the Asset Processor would need to be updated to first locally expand those files.
With regards to nested archives I would be cautious because the experience with nested pak files (previously used by levels) showed it had a significant overhead at runtime. It may be acceptable for source assets, but due to those typically being bulkier than processed assets and there generally being more of them, may also become prohibitively expensive for large projects.
It would help to clarify what the use cases are for having archives for source assets and for nested archives for source assets.

**Reply by nick-l-o3de** [reply]
The asset processor can't actually current cause **new** jobs to be created based on the appearance of intermediate assets.  This is under proposal at the moment, and I like the idea, but I don't want to give the impression that its currently possible.    Builders CAN emit a chain of jobs currently, but they must do so on initial emission.

**Reply by Infurael** [reply]
My bad, I thought this functionality was already available. Probably got it confused with settings up a build chain of jobs that include processing intermediate files.

**Reply by nick-l-o3de** [reply]
Yes.  For example, when a FBX file changes, it could emit a "fbx extract" and then a bunch of "fbx build" job that depends on the "fbx extract job" and thus will wait for it to finish, but currently they all have to be triggered up front by the original source.

**Comment by nemerle** [comment]
> It would help to clarify what the use cases are for having archives for source assets and for nested archives for source assets

I was considering prefab-heavy workloads ( many, many small-ish prefabs ) and efficiency gains of loading them from a simple zip archive, instead of hoping that OS can handle a few thousand files efficiently :)

**Reply by Infurael** [reply]
There are several factors at play I think.
On the editor end it's better to load smartly and minimize the amount of data that actually needs to be loaded into memory. This overall provides a better user experience. Failing that, it would be better for source files to be loaded async through AZ::IO::Streamer and having as many files as possible in flight to maximize parallelization. This in general does a good job of also hiding the overhead of opening/closing files and retrieving file information.
This is a bit more difficult to do for the Asset Builders because they run isolated so there's no centralized file system such as AZ::IO::Streamer. One option would be for the Asset Processor itself to provide this to the Asset Builders, though it would have some overhead due to inter-process communication.
Long term I'm hoping to get to a point where the cache is perpetually creating archives and incrementally compressing the content. In effect the cache would always hold game ready data at that point. Now this doesn't improve the source asset experience, but for intermediate and final assets it would definitely reduce the overhead of opening/close files and retrieving file data, plus the cache would potentially end up being a factor smaller over time so leaving more disk space for content creators.

Right now the overhead for prefabs or other source assets that could potentially be stored in many small files hasn't been shown as a significant issue to the overall asset processing times. The performance problems lie more in things like long times to load fbx files, the time it takes to compress textures or with going through all the shader permutations. Even if files like Prefabs become an issue for the duration of asset processing I'd still be hesitant to put them in archives, mostly because those kind of files benefit from the ability that they can be opened and edited by content creators through basically any text editor. There's also a question around how such an archive would work with source control.

**Reply by nemerle** [reply]
I expect to have a few thousand prefabs strewn over 10-20 directories, so I'm likely to encounter potential slowdowns related to the number of files :smile:  even if a single level only uses prefabs from 2-5 'sets'.

I'd be happy if editor startup time/usage doesn't slow down to a crawl :)

**Comment by nemerle** [comment]
As for intermediate assets, I didn't notice a specific concrete mention of them. 

While one asset processor can save things into the cache and other processors, in turn, can work on those files, this looks to me like a processing graph that saves intermediate steps to the cache directory? To me, the final result of such a graph is 'intermediate asset' 😄

**Comment by nick-l-o3de** [comment]
So the current asset system in O3DE is fairly flat, in that the pipeline is pluggable using builders (which can be in gems) but the builders just transform input data ("Source files") directly to product files (ie, crunched finals) in the cache, which the runtime then uses.     There is no intermediate layer at all.  To be more clear, a builder can emit more than 1 job to process a source, and it can make jobs dependent on each other (so its possible for a job to read the output of another job, in that way be 'intermediate', but more specifically it follows the following rules):

* You must be able to fully regenerate the crunched cached assets given the set of source files, and this generation much be stable (ie, the same set of input source files in your project results in the same outputs in the cache, with eventual consistency, regardless of the order that the files were encountered or processed.  This also means there's no magic secret sauce files that need humans to copy around into the cache or seed projects with.  If you delete your cache and run AP it will rebuild it by running every job again.

* Processing a source file into its products in the cache must not cause *new* source data to be generated that the user then modifies further and checks into source control.  The direction is strictly (from source files to cache) and never (from source files to other source files) or (from cache to source).  More specifically, you don't allow users to *edit* *generated* *files*.  Doing so leads to order-dependent processing, loops in the pipeline, or things like nightly CI pipelines trying to check files into source control, or data loss where a user exports a file that's higher up in the chain and it overwrites user-edited files further down in the chain.  Instead of letting users alter files or checking in files that are generated, you let users alter files that are exclusively designed to store *their edits* - ie, as an overlay to the files or as a set of instructions side by side with the source file that controls how it is processed.

For example, the "Scene Builder" can read a FBX file and produce Atom Meshes, Atom Models, etc.  This processing can be controlled by a side-by-side (filename).scenesettings file.
For example the "Image Builder" can read PNG files and generate an Atom Texture from it.  This process can be controlled by a side-by-side (filename).imagesettings file.   Saving a new version of the image causes the processing to occur again, applying the imagesettings again.

Note that the Builders can perform any number of intermediate _internal_ steps to perform this process and can read any number of other files to achieve it.  So in that model, here are your use cases:

>     Defining asset import pipeline parameters: ( SA->IA ) ( pipeline config default/per asset ):
>         Image format conversion ( texture -> basis etc., color processing )
>         Audio format conversion ( target format selection from a minimal set [wav/ogg]?, ambisonics? )
>         Mesh processing ( optimization, buffer preparation, LOD generation )
  
>   Defining asset finalization pipeline parameters (IA->FA ) ( pipeline config default/per platform/per asset ):
>         Texture downscaling ( format selection, number of mips )
>         Movie resizing/recoding
>         Custom asset packing algorithms ( animation compression, mesh compression, etc. )

In O3DE, you would store the user choices related to these in a side-by-side datafile next to the "source asset" (so for example, blah.png would have blah.png.imagesettings).   The Image Builder would then read both the image source file and the imagesettings file (if available) to decide what to do with that image. THe Image Builder would trigger off *.png but declare imagesettings files, if present, as a source dependency, which causes the job to rerun if that file changes.  Then you'd implement a nice fancy custom dialog for authoring that imagesettings file and controling this pipeline.  You would check the source image file and the imagesettings file into the repository but you would not check the generated files in.

The same approach is currently used for the following:

>       Scene import ( procedural processing, prefab identification - finding commonly used sub-trees, etc. )
The scene system loads a FBX into an intermediate in-memory format, then uses plugins to export it into things like meshes, models, materials, etc.   The loading of it into the intermediate in memory format, as well parameters about what to export and how to do so is stored in the scenesettings file next to the fbx and is authoreed thru the FBX settings dialog.  When the fbx changes, the system reimports it using those settings, so all that is required is re-exporting the FBX from your DCC tool  


>       Custom asset pipelines ( convert csv into textured meshes, other fun stuffs :) )
These would have to be on a case-by-case bassis.


>    DCC live-link ( https://github.com/unity3d-jp/MeshSync )
>        We will likely want two-way interaction, editing an asset in the editor should be reflected back to the DCC tool.

This is where it gets interesting because you have to be extremely careful with your "source of truth" here.  This generally involves a plugin for both the DCC tool and the editor to control how information goes between them and what to transmit.

**Reply by nick-l-o3de** [reply]
NOte that one idea being passed around right now is the concept of "real intermediate assets" but they would still not violate any of those 2 rules.  For example, if you wanted to be able to process a "zip file continaing source assets" (ie, a zip containing texture, fbx, etc) the proposal there is that you'd make a builder that declares it handles *.zip, and it would just unzip it to an intermediate location **but still within the cache or some other git-ignored location that is considered read-only by the user**  you never check those unzipped files into the repository, since the zip would then be the source of truth, and you'd never want the user to modify or author changes to the unzipped files either (As they would get overwritten on re-unzip).   But then the proposal is to have it so that this intermdiate (not-the-cache) location would be monitorable by the usual builders that care about, say, *.fbx and *.png and process the data as usual.

**Reply by nemerle** [reply]
So, 'cache' is a catch-all for:
* place to store imported assets
* temporary work area for asset processor
* place to prepare per-platform unpacked assets

It's a bit messy ( looking at AutomatedTesting/Cache ) right now:
```
all/
all_client/
android/
fence/
ios/
jasper/
mac/
pc/
provo/
salem/  
server/
```
Maybe a slight reorganization of the directory structure in the Cache dir is in order ?
```
editor_runtime/   <- assets for the project that are editor-specific 
work_area/    <- used as a temporary directory by multi-step asset processing ( scanned for changes by all asset_processors )
platforms/ 
    pc/ 
        server/
        full/
    ...
common/   <- platform independent assets
  server/
  full/
```

**Reply by nick-l-o3de** [reply]
Note that the original intent for the cache folder was that it wasn't meant for humans to interact with at all (ie, go look at the cache folder of Unity), or instead contained the pack files already in zip/package format ready to ship.  So it feels like optimizing the readabiity of the cache slaps a band-aid on a larger problem further up the chain in regards to the relationship between intermediates, finals, and so on.

**Reply by nemerle** [reply]
I 100% agree, but right now we don't seem to have a target 'workflow' that would inform any kind of work meant to fix this 😄 

Godot might be a nice example of the Source->Intermediate layout ( nomenclature converted to closer mimic o3de ) :

* Assets are handled by plugins ( think `AssetBuilderDesc::m_patterns` matching files )
* Every plugin creates a 'metadata' file for the asset i.e. `image.png.import`, that contains processing parameters
* processed files are stored in various internal formats in `.import` directory ( AssetProcessor producing multiple texture formats for all known platforms )
* Engine can be extended by plugins that load new internal formats or import them into already recognized internal formats.

Right now we have something a bit similar, although we store parts of the metadata in the local sqlite database and don't seem to have standardized per-asset metadata available for all asset types.

**Reply by nick-l-o3de** [reply]
Yes, I'd love to see a standard metadata API.  I just want to make sure we seperate "stuff users author" (goes into git/perforce) from "stuff generated" (goes into cache or others).   So image.png.metadata -> should go next to image.png      ... but image.atompcdx12 -> goes in cache.  One other consideration is we don't want to generate files into the source folder (Even meta files) during automatic headless operations.  If for example your nightly CI/CD started needing to check things out of perforce or make git commits because its "creating new files" in the repository, this would be an issue.

The one exception to this I could see is meta-files (which, e.g., could contain the identifier of an asset or other default data) but even then we'd have to find a way to signal that CI/CD operations or automated testing does not touch this stuff, only human beings.

**Reply by nemerle** [reply]
The solution to 'generating files' would be to provide the tests with `asset + metadata` pairs ( metadata checked in alongside the asset )
Testing the `metadata` generation step would then be a separate concern.
There is also an orthogonal issue - locating assets ( metadata with UUID + file hash might help out with some problems )

**Reply by nick-l-o3de** [reply]
Yes.  The engine was actually designed with the idea that intermediates would have a meta file with a uuid in it, thus allowing the file to be moved without "changing its identifier" that others refer to it by.   But a lot of systems have started referring to files by their relative path instead of UUID, so there's a lot of plumbing to fix up if we want to go that way.

**Reply by nemerle** [reply]
I think getting back on UUID-track would be a good thing since asset moves and re-linking various dependents is just painful :)

**Reply by nick-l-o3de** [reply]
Yes, it'd be interesting to solicit opinions on this becuase its actually a complicated subject.  Some challenges:
* Many existing systems just use relative path names, including PRefab, and most of Atom.  This means they'd all have to adopt UUID-based asset references instead of (or in addition to) relative path names.  We'd have to make it backward compatible somehow, too.
* scripting (LUA, C++, etc) generally prefers using relative path names, its really inconvenient to type some crazy UUID into your script instead of spawner.Spawn("car.mesh").  But it begs the question why even allow typing in asset names at all, as opposed to e.g. having the script expose a drag and drop field in the property inspector for 'which thing to spawn'.
* people seem to hate metafiles and they make files look cluttered and hard to diff visually.
* UUIDs as identifiers aren't goign to work all the way back into DCC tools so it doesn't quite solve the problem of "I moved a thing that my maya file refers to and now my maya file is broken".  (And I'm not sure its in-scope to fix this, tbh).

We also have some advantages here.  The engine pretends everything has a durable UUID - as in, if you have a c++ class, the correct way to refer to another file is via Asset<T> m_asset; member, which shows up as a drag and drop target in the editor from the asset browser, and also saves it as (UUID, text hint used for human eyes/debugging).  So the engines core system behaves as if UUIDs are a durable identifier for an asset, by default.  Everyone pretends as if there are metafiles and that UUIDs are the correct way to refer to things.   The missing piece here is a place to store the UUID in a durable manner (ie, a metafile), as opposed to computing it (right now the UUID of an asset is the hash of its relative path).

**Reply by nemerle** [reply]
It seems that the start of a solution would be something like an `AssetLocationMapper` that:
* can consume both paths and UUIDs and produce correct asset locations. 
* provide an `asStableLocation(const AZStd::string &)` service that returns an 'opaque' asset location, that can be used to 'port' older assets to a new UUID based location format. ( first implementation can be no-op, and return the argument )
* in the future load the 'mapping' from files ( UUID<->relative_asset_location )

**Reply by nick-l-o3de** [reply]
Yes, we're actually closer to that then you might expect.     The asset processor actualyl already assigns everything a UUID when it first sees them, this is where you'd put a function that generates + stores it in a metafile instead of using the hash of the filepath.  the original UUID could be the hash of the relpath though, so as to not change the numbers and have it work transparently.  
Once it compiles an asset, and stores it in the cache, the AP also updates assetcatalog.xml , which is a map of [uuid] -> [asset info, including its actual relpath in the cache].  Systems are supposed to store the UUID of an asset reference and optionally store its relpath purely for debugging information (since its friendlier to display an error message with the name of an asset instead of a UUID).

But yeah, its all possible.  Its just a matter of getting there. :)

**Reply by HogJonny-AMZN** [reply]
You can also write a custom python asset builder, the example we have of that currently is in the Blast Gem.

**Comment by nemerle** [comment]
The problem I'm having with current o3de's asset setup is likely related to the way I've internalized asset's 'stages of life'

* Source Asset -> an asset produced in a format a DCC tool can efficiently work with
* Intermediate Asset -> processed Source Asset in a format the Editor & Tools can efficiently work with
* Final Asset -> optimized/packed/platform-specific asset that final product runtime can efficiently work with

Our current 'cache' is somehow blurring the boundary between Intermediate and Final assets ( at least to me :) )

**Reply by nick-l-o3de** [reply]
Yes, currently the only kinds of assets O3DE actually cares about in your nomenclature are Intermediate and Final assets.

You put the intermediate assets in your gem / project "assets" folder, it automatically finalizes them into the cache.   It doesn't monitor Source assets at all (ie, it doesn't monitor PSD files, instead it monitors PNG or TIF or JPG, and ie, it doesn't monitor .MA files or .MAX files, it monitors FBX).  It's expected that Tech Art and others would make DCC plugins to auto-export to intermediate (or one click export to intermediate) and handle their sources in a way they desire.  One issue with true sources is that they can be orders of magnitude larger than the intermediate exports.  This means that e.g. a game designer person that wants to be able to build the cache locally only needs the FBX and TIF files, not the original MA files and PSD.  


This is blurred a little further, as some intermediate assets would be really inconvenient for the actual editor and tools to load (for example, a FBX file), so in most cases the editor's visualizations and render use the same rendering tech as the runtime (but embedded in the viewport) which means its loading optimized meshes and stuff from the cache, finals, not intermediates.    This actually makes it a lot simpler.    The only exception to this is when the editor and tools are **editing** things, in which case you're editing things like levels (prefabs / slices) and other authored data directly.  In which case its a source asset that is also an intermediate since there's no part of it thats in a different DCC tool.


HOwever, if we do add support for true intermediate processing, there's no reason why you couldn't shift the pipeline so that the MA files and PSD files are the 'source of truth' and follow the Source -> Intermediate -> Final workflow.  However, you'd have to live with the fact of checking in those huge source files in so that everyone could generate their cache from them.   I think longer term if we want to do that we have to also solve that problem (ie, how can a studio work without everyone needing to get the sources + intermediates, maybe only get the cache + the metadata generated during compile)  None of it is impossible or in fact even that difficult to implement in this engine, its just not there right now.

**Comment by nemerle** [comment]
For anyone interested, I've posted an issue related to this discussion #4457.
